### PR TITLE
Fix tcp/tls socket reusage for multihome opensips

### PR DESCRIPTION
**Pull request by nikbyte**
The patch for https://github.com/OpenSIPS/opensips/issues/2512

### CODE REVIEW

**Comment by nikbyte** [comment]
With this patch when opensips is matching existing tcp connection, it compares source socket and reuse only in case of source sockets are the same.

File: net/net_tcp.c [hunk 1]
```diff
@@ -852,6 +857,8 @@ struct tcp_connection* tcp_conn_create(int sock, union sockaddr_union* su,
 	if (!send2main)
 		return c;
 
+	c->send_sock = si;
```
Comment (razvancrainea) [inline net/net_tcp.c:860]: I don't think there's any need of a new `c->send_sock` in the connection structure, as you already have that information in the `c->rcv.bind_address` field.

File: net/net_tcp.c [hunk 2]
```diff
@@ -461,10 +462,14 @@ int tcp_conn_get(unsigned int id, struct ip_addr* ip, int port,
...
 					a->parent->rcv.src_port);
 				print_ip("ip=",&a->parent->rcv.src_ip,"\n");
+				if (send_sock != NULL && a->parent->send_sock != NULL)
+					print_ip("requered send_sock ip=",&send_sock->address,"\n");
+					print_ip("found send_sock ip=",&a->parent->send_sock->address,"\n");
 #endif
 				c = a->parent;
 				if (c->state != S_CONN_BAD &&
 				    c->flags&F_CONN_INIT &&
+				    send_sock == a->parent->send_sock &&
```
Comment (razvancrainea) [inline net/net_tcp.c:472]: If a socket is not forced (thus `send_sock` is NULL), then we might want to use any connection towards that destination, don't we?

**Review by razvancrainea — CHANGES_REQUESTED** [review]

**Comment by nikbyte** [comment]
@razvancrainea Thank you very much for your useful comments. I fixed regarding your notes.

File: net/net_tcp.c (hunk 2 above)
Comment (nikbyte) [reply net/net_tcp.c]: Fixed

File: net/net_tcp.c (hunk 1 above)
Comment (nikbyte) [reply net/net_tcp.c]: Fixed

**Review by razvancrainea — APPROVED** [review]
This looks all right now, merging it.

File: net/net_tcp.c [hunk 3]
```diff
@@ -461,10 +462,14 @@ int tcp_conn_get(unsigned int id, struct ip_addr* ip, int port,
 					a, a->parent, a->parent->id, a->port,
 					a->parent->rcv.src_port);
 				print_ip("ip=",&a->parent->rcv.src_ip,"\n");
+				if (send_sock != NULL && a->parent->send_sock != NULL)
+					print_ip("requered send_sock ip=",&send_sock->address,"\n");
```
Comment (wdoekes) [inline net/net_tcp.c:466]: Assuming that should be `required`?

Comment (razvancrainea) [reply net/net_tcp.c:466]: This has been fixed before merging, but replaced with `requested` instead of `required`.

**Pull request merged**